Add cassandra option test (#644)
vdaas/vald: docs/contributing/coding-style.md

In Vald, the functional option pattern is widely used in Vald.
You can refer to [this section](#Struct-initialization) for more details of the use case of this pattern.

> We provide the following errors to describe the error to apply the option.
> 
> | Error | Description |
> |----|----|
> | errors.ErrInvalidOption | Error to apply the option, and the error is ignorable |
> | errors.ErrInvalidOptionWithError | Error to apply the option with error return from the library (e.g. time.Parse error), and the error is ignorable |
> | errors.NewErrCriticalOption | Critical error to apply the option, the error cannot be ignore and should be handle |
> | errors.NewErrCriticalOptionWithError | Critical error to apply the option with error return from the library, the error cannot be ignore and should be handle |
> 

Reviewer: these are functions not errors.

```suggestion
We provide the following functions to raise errors when applying the options.

| Function | Description |
|----|----|
| errors.NewErrInvalidOption | Returns error to apply the option, and the error is ignorable |
| errors.NewErrInvalidOptionWithError | Returns error to apply the option with error return from the library (e.g. time.Parse error), and the error is ignorable |
| errors.NewErrCriticalOption | Returns critical error to apply the option, the error cannot be ignore and should be handle |
| errors.NewErrCriticalOptionWithError | Returns critical error to apply the option with error return from the library, the error cannot be ignore and should be handle |

```

| Error | Description |
|----|----|
| errors.ErrInvalidOption | Error to apply the option, and the error is ignorable |

> | errors.ErrCriticalOption | Critical error to apply the option, the error cannot be ignore and should be handle |

Reviewer 2: ```suggestion
| errors.ErrCriticalOption | Critical error to apply the option, the error cannot be ignored and should be handle |
```

}
```

> If the functional option error is a critical error, we should wrap it with `ErrCriticalOption` error and return it to the caller side to decide the next action.

Reviewer 2: ```suggestion
If the functional option error is a critical error, we should wrap it with the `ErrCriticalOption` error and return it to the caller side to decide the next action.
```

> ```go

Reviewer 2: could you please check the preview of markdown?

func WithConnectTimeout(dur string) Option {
    return func(c *client) error {
        if dur == "" {
            return errors.ErrInvalidOption("connectTimeout", dur)
        }
        d, err := timeutil.Parse(dur)
        if err != nil {

>             return errors.NewErrCriticalOption(err)

Reviewer 2: should we wrap `ErrInvalidOption `??
I think we don't notice which option error.:thinking:

Reviewer 3: should we change `NewErrCriticalOption(err error)` to `NewErrCriticalOption(opname string, val interface{}, err error)`

Reviewer 2: @kevindiu 
```
package errors

var (
	// ErrInvalidOption returns invalid option error.
	ErrInvalidOption = func(name string, val interface{}) error {
		if val == nil {
			return Errorf("invalid option. name: %s, val: nil", name)
		}
		return Errorf("invalid option. name: %s, val: %#v", name, val)
	}

	// ErrInvalidOptionWithError --
	ErrInvalidOptionWithError = func(name string, val interface{}, err error) error {
		optErr := ErrInvalidOption(name, val)
		return Wrap(optErr, err.Error())
	}
)

// ErrCriticalOption represent the critical option error
type ErrCriticalOption struct {
	err error
}

func NewErrCriticalOption(name string, val interface{}) error {
	return &ErrCriticalOption{
		err: ErrInvalidOption(name, val),
	}
}

// NewErrCriticalOptionWithError --
func NewErrCriticalOptionWithError(name string, val interface{}, err error) error {
	return &ErrCriticalOption{
		err: ErrInvalidOptionWithError(name, val, err),
	}
}

func (e *ErrCriticalOption) Error() string {
	return Wrap(e.err, "invalid critical option").Error()
}

func IsCriticalOptionError(err error) bool {
	switch err.(type) {
	case *ErrCriticalOption:
		return true
	default:
		return false
	}
}

```

cc: @vankichi

    srv := new(server)
    for _, opt := range opts {
        if err := opt(srv); err != nil {
            werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))

>             log.Warn(werr)

Reviewer 2: I think you should error output when an critical error`s type  is not IsCriticalOptionError
Could you please read this ticket`s exampe?
https://github.com/vdaas/vald/projects/3#card-44665812

Reviewer 3: Sorry maybe I am not quite understand your opinion, can you write example code?

Reviewer 2: ```
w := new(hoge)
for _, opt := range append(defaultOpts, opts...) {
    if errors. IsCriticalOptionError(err) {
        return nil, err
    }

    log.Warnf(err, errors.OptionAllyFailed)
}

```

Reviewer 3: @hlts2 I think your implementation will not log the error message when it is an critical error. 
My implementation will log the error first and return it if it is critical error.

```go
func New(opts ...Option) (Server, error) {
    srv := new(server)
    for _, opt := range opts {
        if err := opt(srv); err != nil {
            werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))
            log.Warn(werr)

            if errors.IsCriticalOptionError(err) {
                return nil, werr
            }
        }
    }
}
```

Do you think it is ok?

Reviewer 3: @hlts2 I think the following is better, what do you think?
Your implementation will print out the logging message without the function name

```go
w := new(hoge)
for _, opt := range append(defaultOpts, opts...) {
    werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))
    if errors. IsCriticalOptionError(err) {
        log.Error(werr)
        return nil, err
    }

    log.Warnf(werr)
}

```

Reviewer 2: it looks good to me

    srv := new(server)
    for _, opt := range opts {
        if err := opt(srv); err != nil {
            werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))

> 

Reviewer 4: ```suggestion
```
```suggestion

```

Reviewer 3: I think it is better to keep this line break to differentiate `werr` and `err`

    srv := new(server)
    for _, opt := range opts {
        if err := opt(srv); err != nil {
            werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))

>             log.Warn(werr)

Reviewer 3: @hlts2 should we log the error first before the error return?
